### PR TITLE
New cd images (ADX support and ARM64)

### PR DESCRIPTION
**Pull request by tarakby**
Update the node software process after the latest crypto module changes: 
 1. the crypto module has non-portable code that requires proper cross compilation when building for a different target
 2. the crypto module uses CPU's ADX instructions. If the target CPU does not support ADX, such code should be disabled. 

In this PR:
- fix a bug in a previous PR ([commit](https://github.com/onflow/flow-go/commit/1d0b8f906de5158f6358cd75543ef582953c9e4d#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R251)) where the empty `netgo` tag was removed. This fixes the build without `netgo`. 
- rename the build that uses the local script arch and ADX detection to "native build". These builds assume the image is built and run on the same machine type (arch and cpu properties). This is the case of CI tests and integration tests for instance. 
- add new images to build production images with and without ADX code:
   - with ADX code: can only run on a machine with ADX support, this is the default case, and it offers optimal performance.
   - without ADX code: uses portable code and can run on all machines. The build requires adding a `CGO` flag, passed as an argument to `Dockerfile`. It offers slower performance. 
- add a new image for `arm64` cross build. The build requires specifying the cross-compiler in `CC`. `CC` is passed to the `Dockerfile` as an argument.  
- add all new images build to `cd.yml` and add a step to install the cross-compilation tools (works only on Debian)
- update `builds.yml` to use the new commands.
- image names in `Makefile` use `IMAGE_TAG` only (which is derived from the short commits in most cases)

These are the images built in `cd.yml` (assumed to run on `amd64`): 
 - with ADX code
 - with ADX code and without `netgo`.
 - without ADX code and without `netgo`.
 - arm64 images
 ( without ADX code and with `netgo` was skipped for now but can be added if needed. With the current PR, if we disable ADX, `netgo` has to be disabled too. This assumes that there are no machines that cannot run images with `netgo` disabled)
`builds.yml` can generate the same images but the arm64 ones, because of the limitation of number of input boxes in githhub workflows.

### CODE REVIEW

File: .github/workflows/cd.yml
Comment (tarakby) [inline .github/workflows/cd.yml:33]: needed for the `arm` cross-build

File: .github/workflows/ci.yml
Comment (tarakby) [inline .github/workflows/ci.yml:190]: for CI runs, native build is fine

File: Makefile
Comment (tarakby) [inline Makefile:285]: native builds detect whether the current machine supports ADX

File: Makefile
Comment (tarakby) [inline Makefile:292]: Optimal case (with ADX and with netgo) - this is called in CD.

Note that the image has the same "tag" as the native builds. Native builds (locally or CI) are not run in CD.

File: Makefile
Comment (tarakby) [inline Makefile:319]: not sure if I should add "native" to the debug images. Are debug images only built natively?

File: Makefile
Comment (peterargue) [inline Makefile:280]: this only works on debian hosts, which may be confusing since we also have a `install-tools` which works on mac. How about renaming to `install-cross-build-tools-debian` or something like that

File: Makefile
Comment (peterargue) [reply Makefile:319]: afaik, they are only used for localnet/integration test.

File: Makefile
Comment (peterargue) [reply Makefile:292]: so the difference between this an `docker-native-build-collection` is that native allows overriding ADX while this doesn't?

File: Makefile
Comment (tarakby) [reply Makefile]: Great point! I updated an image name and extended the test https://github.com/onflow/flow-go/pull/5254/commits/de0c98c4f0263cabd217e2e2ca4915c5ec96b426

File: Makefile
Comment (peterargue) [inline Makefile:24]: Unfortunately, we had to revert the change that added `+` here because it's not compliant with docker build tags. https://github.com/onflow/flow-go/pull/5267

Let's use `-` for now. We can circle back later to find a way to handle this

```suggestion
IMAGE_TAG_NO_ADX := $(IMAGE_TAG)-without-adx
IMAGE_TAG_NO_NETGO_NO_ADX := $(IMAGE_TAG)-without-netgo-without-adx
IMAGE_TAG_ARM := $(IMAGE_TAG)-arm
```

File: Makefile
Comment (tarakby) [reply Makefile:280]: Ideally we would add an os check within the target, beyond the name. I'll try to add one

File: Makefile
Comment (tarakby) [reply Makefile:292]: This target hardcodes the arguments for a specific target (ideally I would want to hardcode `GOARCH` too, but left it in case we want to call it on `ARM` machines too). 

`docker-native-build-...` set the arguments based on the host (detect ADX support, set `ARCH`)

File: Makefile
Comment (tarakby) [reply Makefile:319]: Great, in that case I may rename the `debug` images to add `native`

**Comment by sjonpaulbrown** [comment]
@tarakby, using this branch, could you please run all of the workflows being modified so that we can verify the execution?

Also, as a general comment, I think we will need to update the following lines in the build workflow to use the new `make` targets

* https://github.com/onflow/flow-go/blob/master/.github/workflows/builds.yml#L119
* https://github.com/onflow/flow-go/blob/master/.github/workflows/builds.yml#L127

**Comment by sjonpaulbrown** [comment]
If the targets are changing, the[ following workflow](https://github.com/dapperlabs/flow-go/blob/master-private/.github/workflows/bn2-create-network.yml#L227-L232) will also need to be updated to align with the target changes.

This will need to be updated separately, but this will break BN2 until the workflow is updated.

File: Makefile
Comment (tarakby) [reply Makefile:280]: https://github.com/onflow/flow-go/pull/5254/commits/0324c18f1fb162450cdf444471a40ad511dd7c58

**Comment by tarakby** [comment]
Thanks @sjonpaulbrown, good points

- I added the new builds in https://github.com/onflow/flow-go/pull/5254/commits/5742af3aa8c41ba6d40bae0199033cd50715b528
- I will update the dapperlabs repo for BN2 once this one is merged 👌🏼

File: Makefile
Comment (tarakby) [reply Makefile:322]: It look like I have to add "native" anyway because of this : https://github.com/onflow/flow-go/blob/master/integration/testnet/container.go#L124-L136

Comment (tarakby) [reply Makefile:322]: https://github.com/onflow/flow-go/pull/5254/commits/097be893ae31c774c300b0a462c5c14e85c6bef3

File: Makefile
Comment (tarakby) [reply Makefile:295]: For better clarity and consistency, I hardcoded `amd64` for production targets that are not natively built and not meant for `arm` https://github.com/onflow/flow-go/pull/5254/commits/d3b99bd42237df016693a1be7aecf2245ab9c744 (though I can revert it if needed)

File: .github/workflows/builds.yml
Comment (peterargue) [inline .github/workflows/builds.yml:41]: I don't think we need a variable for this. It should be the default and built every time

File: .github/workflows/builds.yml
Comment (peterargue) [inline .github/workflows/builds.yml:46]: how about renaming this to
```suggestion
      include_alternative_builds:
```

then when set, it build all alternative builds (`without_adx`, `without_netgo_without_adx`, `arm`)

**Review by peterargue — APPROVED** [review]
Thanks for adding support for ARM builds and unravelling our complex mess of a makefile.

I'm a little confused by all make target names, but I think the changes are correct, so I don't want to hold this up this PR over naming.

We should circle back and refactor the build system soon. It's become pretty complex.

File: .github/workflows/builds.yml
Comment (tarakby) [reply .github/workflows/builds.yml]: implemented in https://github.com/onflow/flow-go/pull/5254/commits/f1e015e30d3ca9975c0bac755dc5f46e9263b3e8 and https://github.com/onflow/flow-go/pull/5254/commits/a3951090c1f53ea324f229d04f623c80ff42ea95

File: .github/workflows/builds.yml
Comment (tarakby) [reply .github/workflows/builds.yml]: Good idea, implemented in implemented in https://github.com/onflow/flow-go/commit/f1e015e30d3ca9975c0bac755dc5f46e9263b3e8

**Comment by tarakby** [comment]
@sjonpaulbrown, the PR is now ready. 

 - `cd.yml` was tested in this [workflow](https://github.com/onflow/flow-go/actions/runs/7631880625/job/20790840384)
 - `builds.yml` was tested in this [workflow](https://github.com/onflow/flow-go/actions/runs/7633959946)

**Comment by tarakby** [comment]
> I'm a little confused by all make target names, but I think the changes are correct, so I don't want to hold this up this PR over naming.
We should circle back and refactor the build system soon. It's become pretty complex.

Agree, the build system could be made simpler.

File: Makefile
Comment (sjonpaulbrown) [inline Makefile:22]: Altering this tag name could impact partner automation. We will want to work with @vishalchangrani to ensure that the partners who rely on this know the image tag is changing.

**Pull request merged**
